File: minacoin-server/miner/classes/Miner.js
'use strict';

const LOG_TAG = 'MNR';

const ioc = require('../../util/iocContainer');
const { Transaction, Wallet } = require('../../wallet');

const logger = ioc.loggerFactory.createLogger(LOG_TAG);
const exception = ioc.ehFactory.createHandler(logger);

/**
 * minacoin: Miner
 * ---------------
 * handles mining new coins 
 *
 * author: John R. Kosinski
 */
class Miner {
    get blockchain() { return this._blockchain; }
    get transactionPool() { return this._transactionPool; }
    get wallet() { return this._wallet; }
    get p2pServer() { return this._p2pServer; }

    /**
     * constructor 
     * @param {Blockchain} blockchain 
     * @param {TransactionPool} transactionPool 
     * @param {Wallet} wallet 
     * @param {IP2PServer} p2pServer 
     */
    constructor(blockchain, transactionPool, wallet, p2pServer) {
        this._blockchain = blockchain;
        this._transactionPool = transactionPool;
        this._wallet = wallet;
        this._p2pServer = p2pServer;
    }

    /**
     * selects transactions from the transaction pool to add to a new block, creates 
     * & adds the new block to the blockchain, and returns the new block. 
     * @returns {Block}
     */
    /*Block*/ mine() {
        return exception.try(() => {
            //get transactions from transaction pool
            const validTransactions = this.transactionPool.validTransactions();

            //add a reward for self
            validTransactions.push(Transaction.rewardTransaction(this.wallet, Wallet.blockchainWallet()));

            //add them into a block
            const block = this.blockchain.addBlock(validTransactions);

            //sync the chain
            this.p2pServer.syncChain();

            //clear the transaction pool
            this.transactionPool.clear();

            //broadcast directive to clear transaction pool
            this.p2pServer.broadcastClearTransactions();

            return block;
        });
    }
}

module.exports = Miner;

File: minacoin-core/src/miner/miner.cpp
#include "miner.hpp" 
#include <vector>

namespace minacoin::miner {
    Miner::Miner(Blockchain* blockchain, Wallet* wallet, TxPool* txPool) {
        this->_blockchain = blockchain;
        this->_wallet = wallet;
        this->_txPool = txPool;
        
        this->initLogger("MINR");
    }
    
    Miner::~Miner() {
        
    }
    
    Block* Miner::mine() {
        this->logger()->info("mining..."); 
        
        auto validTxs = this->_txPool->validTxs(this->_blockchain); 
        
        if (validTxs.size() > 0) {
            
            //add a reward for self 
            auto blockchainWallet = Wallet::blockchainWallet(); 
            auto rewardTx = Transaction::reward(this->_wallet->address(), blockchainWallet->address()); 
            blockchainWallet->signTransaction(rewardTx);
            validTxs.push_back(rewardTx);
            
            //add them into a block
            vector<IBlockDataItem*> blockData(validTxs.begin(), validTxs.end());
            auto block = this->_blockchain->addBlock(blockData); 
            
            if (block) {
                //clear tx pool 
                this->logger()->info("block %s mined successfully; clearing transaction pool", block->hash().c_str());
                this->_txPool->clear(); 
            }
            else {
                //clear the transaction pool; maybe we have an old or corrupt pool
                this->logger()->warn("mined block is invalid; clearing transaction pool");
                this->_txPool->clear();
            }
            
            return block;
        }
        else {
            this->logger()->info("no valid transactions to mine");
        }
        
        return nullptr;
    }
}